### PR TITLE
Support per-parameter test decoration

### LINKED ISSUE #79161: Support more granular test decoration
Currently, the `DecorateInfo` class only supports conditionally applying decorators to test functions based on test class name, base test name, device, and dtype. Example usage:

https://github.com/pytorch/pytorch/blob/7cb4a76844edfed7e31c50ec5af8fbfe8cbb6ca0/torch/testing/_internal/common_methods_invocations.py#L9800-L9805

It would be useful to support more granular test decoration, ideally based on any parameter that is passed to the test function rather than just `device` / `dtype`.

Case in point: #78735 is adding support for running tests across train / eval modes for `ModuleInfo`. Some tests are expected to fail for eval mode only, and there's no granular way to indicate this.

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91658

Continuation of #79979.

Fixes #79161

This PR does the following:
* Expands the `parametrize_fn()` signature from returning a 3-tuple of `(test, test_name, param_kwargs)` to returning a 4-tuple of `(test, test_name, param_kwargs, decorator_fn)`. Expected signature for the addition is `decorator_fn(param_kwargs) -> List[decorator]` i.e. given the full set of test params, return a list of decorators to apply.
    * `modules`, `ops`, and `parametrize` now fit the new signature, returning `decorator_fn`s instead of applying decorators themselves.
    * `instantiate_parametrized_tests()` and `instantiate_device_type_tests()` now call the returned `decorator_fn`, passing in the full set of `param_kwargs` (after composition + `device` / `dtype` additions) and applying the returned decorators.
    * Composing multiple `parametrize_fn`s also composes the corresponding `decorator_fn`s; the composed `decorator_fn` simply concatenates the decorator lists returned by the constituents.
* Expands `DecorateInfo.is_active` to support callables:
```python
DecorateInfo(
    unittest.expectedFailure, "TestOps", "test_python_ref_executor",
    device_type='cuda', active_if=lambda params: params['executor'] == 'nvfuser'
),
```
* Adds several tests to `test/test_testing.py` ensuring proper decoration using `@parametrize`, `@modules`, and `@ops`.
* (minor) Fixes a couple `ModuleInfo` naming oddities uncovered during testing.

### CODE REVIEW

File: torch/testing/_internal/common_device_type.py [hunk 1]
```diff
@@ -271,9 +271,17 @@ def _dtype_test_suffix(dtypes):
 
 def _update_param_kwargs(param_kwargs, name, value):
     """ Adds a kwarg with the specified name and value to the param_kwargs dict. """
+    # Make name plural (e.g. devices / dtypes) if the value is composite.
+    plural_name = '{}s'.format(name)
```
Comment (malfet) [inline torch/testing/_internal/common_device_type.py:275]: Or `f'{name}s'`
```suggestion
    plural_name = name + 's'
```

**Review by malfet — APPROVED** [review]
LGTM, though I wonder how much it will slow down the testing in general.

**Comment by jbschlosser** [comment]
> LGTM, though I wonder how much it will slow down the testing in general.

Good point; I did do some smoke testing with a few test files and didn't notice any non-negligible slowdowns
